### PR TITLE
initial SUSE support

### PR DESCRIPTION
**Pull request by nevion**
This is a WIP and was requested to help debug it's development

### CODE REVIEW

**Comment by redhatrises** [comment]
Hello,

Thank you for the patch! A couple of comments:

- Please comment out the `SUSE12` and `SUSE12-dist` in the `Makefile` for now until the content is more stable as this breaks the Jenkins builds.
- Please change references from just `SUSE` to `Novell SUSE`.
- Please remove ` SUSE/12/input/auxiliary/stig_overlay.xml` as Novell SUSE does not have a STIG.
- Please change all `<platform>SUSE 12</platform>` to `<platform>Novell SUSE Enterprise Linux 12</platform>`
- More comments inline.

File: SUSE/12/input/oval/templates/template_BASH_package_installed [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# platform = SUSE Enterprise 12
```
Comment (redhatrises) [inline SUSE/12/input/oval/templates/template_BASH_package_installed]: This should be Novell SUSE Enterprise Linux 12 for all references.

File: SUSE/12/input/profiles/rht-ccp.xml [hunk 2]
```diff
@@ -0,0 +1,140 @@
+<Profile id="rht-ccp" xmlns="http://checklists.nist.gov/xccdf/1.1">
```
Comment (redhatrises) [inline SUSE/12/input/profiles/rht-ccp.xml]: Remove this file  `SUSE/12/input/profiles/rht-ccp.xml`

File: SUSE/12/input/oval/platform/suse12-cpe-dictionary.xml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<cpe-list xmlns="http://cpe.mitre.org/dictionary/2.0"
+          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+          xsi:schemaLocation="http://cpe.mitre.org/dictionary/2.0 http://cpe.mitre.org/files/cpe-dictionary_2.1.xsd">
+          <cpe-item name="cpe:/o:suse:sles:12">
```
Comment (redhatrises) [inline SUSE/12/input/oval/platform/suse12-cpe-dictionary.xml]: This should actually be `o:novell:suse_linux_enterprise_server:12`

File: SUSE/12/input/oval/platform/suse12-cpe-dictionary.xml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+            <title xml:lang="en-us">SUSE Enterprise Server 12</title>
+            <!-- the check references an OVAL file that contains an inventory definition -->
+            <check system="http://oval.mitre.org/XMLSchema/oval-definitions-5" href="filename">installed_OS_is_suse12</check>
+          </cpe-item>
+          <cpe-item name="cpe:/o:suse:sled:12">
```
Comment (redhatrises) [inline SUSE/12/input/oval/platform/suse12-cpe-dictionary.xml]: This should actually be `o:novell:suse_linux_enterprise_desktop:12`

File: SUSE/12/input/oval/platform/suse12-cpe-dictionary.xml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+            <title xml:lang="en-us">SUSE Enterprise Desktop 12</title>
+            <!-- the check references an OVAL file that contains an inventory definition -->
+            <check system="http://oval.mitre.org/XMLSchema/oval-definitions-5" href="filename">installed_OS_is_suse12</check>
+          </cpe-item>
+          <cpe-item name="cpe:/o:opensuse:opensuse:42.1">
```
Comment (redhatrises) [inline SUSE/12/input/oval/platform/suse12-cpe-dictionary.xml]: This should be `o:novell:leap:42.1`

File: SUSE/12/input/guide.xml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+creators of this guidance assume no responsibility whatsoever for its use by
+other parties, and makes no guarantees, expressed or implied, about its
+quality, reliability, or any other characteristic.</notice>
+
+<front-matter>The SCAP Security Guide Project<br/>https://fedorahosted.org/scap-security-guide</front-matter>
+<rear-matter>Red Hat and Red Hat Enterprise Linux are either registered
+trademarks or trademarks of Red Hat, Inc. in the United States and other
+countries. All other names are registered trademarks or trademarks of their
+respective companies.</rear-matter>
+<platform idref="cpe:/o:suse:sles:12" />
```
Comment (redhatrises) [inline SUSE/12/input/guide.xml]: This should actually be `<platform idref="cpe:/o:novell:suse_linux_enterprise_server:12" />`
and the following need to be added:
`<platform idref="cpe:/o:novell:suse_linux_enterprise_desktop:12" />`
`<platform idref="cpe:/o:novell:leap:42.1" />`

**Comment by iankko** [comment]
@openscap-jenkins 
[test this please]

**Comment by redhatrises** [comment]
Nack on commit 61312a58c. Linkat (http://linkat.xtec.cat/portal/index.php) is considered a derivative of SuSE not SuSE itself. Derivatives are projects that have spun off or branched off of the main products which in this case is openSUSE, SUSE Server, SUSE Desktop.....

**Comment by redhatrises** [comment]
Also all content related to `firewalld` should be replaced with `SuSEfirewall2`. So in this example, the `package_firewalld_installed.xml` OVAL file should be `package_susefirewall2_installed.xml`.

**Comment by nevion** [comment]
@redhatrises re: derivatives, all SUSE products are now derivatives of SLES. Everything is the same to my knowledge other than repos and some newer packages.  Maybe this isn't the notion of derivatives you had originally conceived but the lines on SUSE are more blurry and compatible than everything else so far, and do not appear to warrant the effort/copying to maintain the split..

**Comment by redhatrises** [comment]
@nevion that is why they should all be from the same content as they are more blurred. Derivatives are for a clear demarcation between projects. Because of the blur, all SuSE products can be in the same content rather than generating splits, copying, and derivatives.

**Comment by nevion** [comment]
@redhatrises  so what are the practical steps to handle this?

On a related note, all the files under /etc/pam.d are different paths on SUSE - how to override the handling of this while  not copying n times and in compliance with above.  I was not able to figure out how to modify the files under shared in a way that didn't use regexp paths matching (which I supposed would not be liked).

**Comment by redhatrises** [comment]
@redhatrises  so what are the practical steps to handle this?

Using `multi_platform_suse` (which would contain SLES, SLED, etc.) and having separate XCCDF profiles.

> On a related note, all the files under /etc/pam.d are different paths on SUSE - how to override the handling of this while not copying n times and in compliance with above. I was not able to figure out how to modify the files under shared in a way that didn't use regexp paths matching (which I supposed would not be liked).

As long as it doesn't conflict (make something fail when it originally passed), this shouldn't be an issue. However for clarity, can you provide an example of what you are referring to in the pam files?

**Comment by nevion** [comment]
@redhatrises I am using multi_platform_suse already (just refreshed push to reflect) , wouldn't each xccdf profile for these three be _exactly_ the same though?

re pam files for instance,
`    <ind:filepath>/etc/pam.d/system-auth</ind:filepath>`
to
`    <ind:filepath operation="pattern_match">/etc/pam.d/system-auth|/etc/pam.d/common-password</ind:filepath>`
This opens the test up to error due to platform difference.  But if we don't do it here, then we must make and maintain redundant test implementations who only differ in which file they check.

**Comment by redhatrises** [comment]
> wouldn't each xccdf profile for these three be exactly the same though?

Yes. But one of them can contain all the checks. See https://github.com/OpenSCAP/scap-security-guide/blob/master/RHEL/7/input/profiles/stig-rhel7-workstation-upstream.xml for an example.

> This opens the test up to error due to platform difference. But if we don't do it here, then we must make and maintain redundant test implementations who only differ in which file they check.

That may be fine and a very valid thing to do in some cases. I don't think that in this scenario that it would be much of an issue. From a regex perspective, it would be simpler to write 
`<ind:filepath operation="pattern_match">/etc/pam.d/(system-auth|common-password)</ind:filepath>`

**Comment by nevion** [comment]
@redhatrises it's still not clear to me what you propose in terms of minimizing copy/paste, can you fork my repo and make some of the proposed changes to how the distributions are handled based? 

I also want to say this is WIP - please do not merge this until I explicitly ask, even if it passes all your criteria.  I'm using the PR strategy to help aid writing these files only.

A few earlier items:

> Please comment out the SUSE12 and SUSE12-dist in the Makefile for now until the content is more stable as this breaks the Jenkins builds.

Still? 

>Please remove SUSE/12/input/auxiliary/stig_overlay.xml as Novell SUSE does not have a STIG.

How does one come into being? At the moment it seems far more useful to have that file until the vendor (Novel?) gets one generated.  Is there a compromise of sort to work out?  The entire STIG seems to apply.

**Comment by redhatrises** [comment]
> it's still not clear to me what you propose in terms of minimizing copy/paste, can you fork my repo and make some of the proposed changes to how the distributions are handled based? 

Will do. 

> I also want to say this is WIP - please do not merge this until I explicitly ask, even if it passes all your criteria. I'm using the PR strategy to help aid writing these files only.

One problem with this is that there are now 800 files to review! That number is only increasing. The smaller the PRs the better.

>>Please remove SUSE/12/input/auxiliary/stig_overlay.xml as Novell SUSE does not have a STIG.

>How does one come into being? At the moment it seems far more useful to have that file until the vendor (Novell?) gets one generated.  Is there a compromise of sort to work out?  The entire STIG seems to apply.

The stig_overlay.xml file is taken from an actual STIG document which is DISA generated for a specific product type, so using the Red Hat stig_overlay.xml for SUSE is a no-go from the start.

> But that isn't what's in /etc/os-release for SLES or opensuse (the leap CPE you had). Do these not need to match up?

No. See here: https://web.nvd.nist.gov/view/cpe/search/results?keyword=opensuse&status=FINAL&orderBy=CPEURI&namingFormat=2.3

**Comment by redhatrises** [comment]
@nevion 

Looking into this further, openSUSE will need its own content separate from SLE. Reason for this is because openSUSE has a faster release cadence (technically considered upstream of SLE), new packages and implementations, is community based which most likely will not go through security certifications due to the high cost to get an OS certified. Because of this, things like stigids (assigned to a specific OS and OS version) and cces (assigned to a specific OS and OS version) would not apply to openSUSE. openSUSE CPEs will also be changing frequently as well as the security posture of the system will most likely be changing frequently as well.

Novell (Micro Focus), on the other hand, goes through the required certification process, so SLE would get its own stigids and cces which are different from RHEL. SLES and SLED will be able to use the same content tree which will help to minimize the amount of SLE content to generate. SLE CPEs will be added based on new releases and the frequency of changing (including security changes) will be slower compared to openSUSE.

See the Fedora vs. RHEL content which is the similar to openSUSE vs SLE.

I have added a branch with your work here: https://github.com/redhatrises/scap-security-guide/tree/initial_suse_support and will be making the proposed changes I have been talking about in that branch for you to look into.

**Comment by nevion** [comment]
@redhatrises  I figured you'd recommend put it's own content directory.  I'm not sure if it's technically going to change result but the chance for repeated work is high though because as of 42.1, they want to track SLE such that it just works going from openSUSE to SLES.  They recognized software development is easier on openSUSE and want to encourage a streamlined adoption to their enterprise software for the deployment phase.  This is a good part of why openSUSE versioning went from 13 to 42 and introducing the term word Leap.  Personally having used SUSE and SLES for a ~10 years, I'm not aware of any differences besides less repositories and software availability.  All the documentation, tools, configurations and directories have been the same.  https://en.opensuse.org/Portal:Leap  Unlike redhat, I don't think it will end up as much of testing grounds like Fedora vs RHEL, although there's been some acknowledgement that that will happen to a degree, both ways. 

Anyway, I am quite sure 42.x will track 12.x closely, with the same cadence, but with openSUSE trailing behind in release date, and I do want to seek "re-usage" of the certifications / stigs if that's at all possible to support software development prior to deployment, when that is good enough.  I haven't torn very heavily into the Fedora vs RHEL much yet because the tree is very out of date and at least earlier I knew even less about this system of tests assembles/works.  I'll give it a deeper gander.

**Pull request closed without merging**